crates/bevy_ecs/src/component/mod.rs

```diff
@@ -306,10 +307,10 @@ impl Components {
     }
 }
 
-#[derive(Copy, Clone, Debug)]
+#[derive(Clone, Debug)]
 pub struct ComponentTicks {
     pub(crate) added: u32,
-    pub(crate) changed: u32,
+    pub(crate) changed: Cell<u32>,
```

Comment:
I think this is actually counterintuitively a reduction in safety. `Cell` is `!Sync`, and calling `set_changed()` from multiple threads at the same time violates that constraint. On `main` we only provide mutable access to ComponentTicks when the caller has unique access to the corresponding component across all threads, so the old UnsafeCell wrapper doesn't allow invalid access in userspace. But now callers without unique access can call `set_changed`, which allows cross thread mutation of `Cell`.

Reply:
I think technically userspace is still "safe" right now with this pr because there are no public interfaces that expose `&ComponentTicks` directly from (ex: `Res<T>` has an internal `&ComponentTicks` reference). But imo this change makes it much easier to accidentally expose unsafe behavior. Its tempting to look at `Res<T>` and think "we should expose `set_changed()` here ... its a safe api so we're fine".

Reply:
I see. I somehow ignored the part where things get sent across threads, and compiler let me do it because of existing pointer casts. I'll try to work around that and get rid of Cell. I still think I can end up with better placement for `UnsafeCell`, as I believe having it over the whole `Vec` wasn't correct. Safe alternative would be to use an atomic in place of cell, but It's likely to impact performance, so I'm not doing it right now.

I think at some point we would also think about using [`GhostCell`](http://plv.mpi-sws.org/rustbelt/ghostcell/) once its implementation matures. I believe we would be able to remove a lot of unsafety that way.

Reply:
Cool cool. Definitely happy to consider alternatives. For now, can you revert to the current UnsafeCell impl so we don't need to block the drop fixes on UnsafeCell changes (which appear to be orthogonal).

Reply:
I don't want to revert it outright, as I believe a lot of those changes are still valid and benefitial. I will make it more similar to previous implementation though and will keep `UnsafeCell` around.

One reason is that I believe this line inside `get_with_ticks` is a sneaky UB, and I want to get rid of it
```rust
let ticks = &mut *self.ticks.get();
```
Because even though we only really use that reference in order to get the ticks at specific index, we do create a unique reference to the vector while possibly holding unique references to its parts (some ticks at other indices). It's totally possible to reallocate that vec by accident through shared reference and invalidate all the live ticks references, which should not ever be possible.

For that reason, I want to use `Vec<UnsafeCell<ComponentTicks>>` instead.